lib/ret_web/channels/hub_channel.ex

```diff
@@ -145,22 +159,129 @@ defmodule RetWeb.HubChannel do
+
+        template |> String.ends_with?("-pen") ->
+          account |> can?(spawn_drawing(hub))
+
+        true ->
+          # We want to forbid messages if they fall through the above list of template suffixes
+          false
+      end
+
+    if should_broadcast do
+      data =
+        payload["data"]
+        |> Map.put("creator", socket.assigns.session_id)
+        |> Map.put("owner", socket.assigns.session_id)
+
+      payload = payload |> Map.put("data", data)
+
+      socket = socket |> maybe_store_created_object(payload)
+
+      broadcast_from!(socket, event, payload)
+
+      {:noreply, socket}
+    else
+      {:noreply, socket}
+    end
+  end
+
+  # Captures inbound NAF update messages
+  def handle_in("naf" = event, %{"dataType" => "u"} = payload, socket) do
+    if payload["data"] |> should_broadcast_object_manipulation(:update, socket) do
+      broadcast_from!(socket, event, payload)
+    end
+
+    {:noreply, socket}
+  end
+
+  # Captures inbound NAF multi update messages
+  def handle_in("naf" = event, %{"dataType" => "um"} = payload, socket) do
+    %{"data" => %{"d" => updates}} = payload
+
+    filtered_updates = updates |> Enum.filter(&(&1 |> should_broadcast_object_manipulation(:update, socket)))
+
+    if filtered_updates |> length > 0 do
+      payload = payload |> Map.put("data", payload["data"] |> Map.put("d", filtered_updates))
+      broadcast_from!(socket, event, payload)
+    end
+
+    {:noreply, socket}
+  end
+
+  # Captures inbound NAF removal messages
+  def handle_in("naf" = event, %{"dataType" => "r"} = payload, socket) do
+    if payload["data"] |> should_broadcast_object_manipulation(:remove, socket) do
+      broadcast_from!(socket, event, payload)
+    end
+
+    {:noreply, socket}
+  end
+
+  # Captures inbound NAF drawing buffer updates
+  # Drawings are special since we implemented our own networking code, their data type is actually an identifier
+  # for a particular instance of a networked drawing.
+  def handle_in("naf" = event, %{"dataType" => "drawing-" <> drawing_network_id} = payload, socket) do
+    account = Guardian.Phoenix.Socket.current_resource(socket)
+    hub = socket |> hub_for_socket
+
+    created_object = socket.assigns.created_objects |> Enum.find(&(&1.network_id == drawing_network_id))
+
+    # If created_object is nil, we've received a message for a drawing that has not received a first sync yet,
+    # or was denied creation. so just ignore it.
+    if created_object != nil do
+      is_creator = created_object.creator == socket.assigns.session_id
+
+      if is_creator or account |> can?(spawn_drawing(hub)) do
+        broadcast_from!(socket, event, payload)
+      end
+    end
 
-    payload = payload |> Map.put("data", data)
-    broadcast_from!(socket, event, payload)
     {:noreply, socket}
   end
 
+  # Fallthrough for all other dataTypes
   def handle_in("naf" = event, payload, socket) do
     broadcast_from!(socket, event, payload)
```

Comment:
what ends up going here? perhaps we should rip this out (and whitelist if needed) to minimize fallthrough risk if there are only a few places?